shiny/ui/_theme_brand.py

```diff
@@ -0,0 +1,528 @@
+                mapped[sass_var] = thm_color
+
+        brand_color_palette = self.brand.color.to_dict(include="palette")
+
+        # Map the brand color palette to Bootstrap's named colors, e.g. $red, $blue.
+        for pal_name, pal_color in brand_color_palette.items():
+            if pal_name in bootstrap_colors:
+                mapped[pal_name] = pal_color
+
+            # Create Sass and CSS variables for the brand color palette
+            color_var = sanitize_sass_var_name(pal_name)
+
+            # => Sass var: `$brand-{name}: {value}`
+            brand_sass_vars.update({f"brand-{color_var}": pal_color})
+            # => CSS var: `--brand-{name}: {value}`
+            brand_css_vars.append(f"--brand-{color_var}: {pal_color};")
+
+        return {**brand_sass_vars, **mapped}, brand_css_vars
+
+    def _prepare_typography_vars(self) -> dict[str, str]:
+        """Typography: Create a list of Bootstrap Sass variables"""
+        mapped: dict[str, str] = {}
+
+        if not self.brand.typography:
+            return mapped
+
+        brand_typography = self.brand.typography.model_dump(
+            exclude={"fonts"},
+            exclude_none=True,
+        )
+
+        for field, prop in brand_typography.items():
+            if field not in typography_map:
+                self._handle_unmapped_variable(f"typography.{field}")
+                continue
+
+            for prop_key, prop_value in prop.items():
+                if prop_key in typography_map[field]:
+                    if field == "base" and prop_key == "size":
+                        prop_value = str(maybe_convert_font_size_to_rem(prop_value))
+
+                    typo_sass_vars = typography_map[field][prop_key]
+                    for typo_sass_var in typo_sass_vars:
+                        mapped[typo_sass_var] = prop_value
+                else:
+                    self._handle_unmapped_variable(f"typography.{field}.{prop_key}")
+
+        return mapped
+
+    def _add_sass_ensure_variables(self):
+        """Ensure the variables we create to augment Bootstrap's variables exist"""
+        self.add_defaults(
+            **{
+                "code-font-weight": None,
+                "code-inline-font-weight": None,
+                "code-inline-font-size": None,
+                "code-block-font-weight": None,
+                "code-block-font-size": None,
+                "code-block-line-height": None,
+                "link-bg": None,
+                "link-weight": None,
+            }
+        )
+        self.add_defaults("// *---- brand: added variables ---* //")
+
+    def _add_sass_brand_grays(self):
+        """
+        Adds functions and defaults to handle creating a gray scale palette from the
+        brand color palette, or the brand's foreground/background colors.
+        """
+        self.add_functions(
+            """
+            @function brand-choose-white-black($foreground, $background) {
```

Comment:
The Sass is getting complex enough now that I think we should consider importing a Sass file?

Reply:
It's not a bad idea but not now; we'd have to implement the Sass layer spec which would be a whole side quest.